### PR TITLE
MOB-1751: Add shared cross-chain payment URI parser

### PR DESCRIPTION
**Pull request by zcash-harry**
## Summary

- add a typed payment_uri crate for Bitcoin, Litecoin, Ethereum, and Solana payment requests
- delegate Ethereum parsing and categorization to the existing eip681 crate without changing its parser logic
- expose one versioned binding representation for every supported scheme
- validate Bitcoin and Litecoin addresses, Solana public keys, exact decimal amounts, duplicate parameters, and required extensions
- keep wallet asset allowlisting and token precision checks in consuming apps

Bitcoin implements the on-chain subset of BIP 321. Litecoin follows Litecoin Core compatible BIP-21 behavior. Solana follows the Solana Pay transfer and transaction-request specification.

NEAR is intentionally excluded because no normative NEAR payment URI standard was found.

## Testing

- cargo test -p payment_uri --all-features
- cargo check -p payment_uri --no-default-features
- cargo clippy -p payment_uri --all-features --all-targets -- -D warnings
- RUSTDOCFLAGS=-D warnings cargo doc -p payment_uri --all-features --no-deps

Authored with OpenAI Codex.

### CODE REVIEW

**Comment by zcash-harry** [comment]
Should btc and sol have it's own crate like eip-681, in the future?

File: components/payment_uri/src/bitcoin.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    for parameter in query.into_iter().flat_map(|query| query.split('&')) {
+        let (raw_key, raw_value) = parameter.split_once('=').unwrap_or((parameter, ""));
+        let normalized = match currency {
+            Currency::Bitcoin => raw_key.to_ascii_lowercase(),
+            Currency::Litecoin => raw_key.to_owned(),
+        };
+        let (required, key) = normalized
+            .strip_prefix("req-")
+            .map_or((false, normalized.as_str()), |key| (true, key));
```
Comment (nuttycom) [inline components/payment_uri/src/bitcoin.rs:157]: Interesting; does the bitcoin payment URI standard use a `req-` prefix like ZIP 321 does?

File: components/payment_uri/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+#[cfg(feature = "json")]
+fn utxo_json(request_type: &str, request: &UtxoPaymentRequest) -> Value {
+    json!({
+        "version": JSON_VERSION,
+        "type": request_type,
+        "address": request.address(),
+        "network": match request.network() {
+            Network::Mainnet => "mainnet",
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs]: Very nitpick: do we have a `as_str()` on the Network type?

File: components/payment_uri/tests/parse.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
+use payment_uri::{Network, PaymentRequest, SolanaRequest};
+
+#[test]
+fn parses_bitcoin_request_and_rejects_unsafe_extensions() {
+    let request = PaymentRequest::parse(
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs:13]: Very nit: this test does two things: check a correct value, and check multiple invalid, i.e. positive and negative tests.
The valid version/positive could be in one test so we can include more valid versions (and maybe use some proptest if possible).
The invalid tests could go either go in one single test, or go in multiple tests whose name includes the invalid path. Also here proptest could be used, and AI could be asked to generate more invalid scheme.

File: components/payment_uri/src/bitcoin.rs (hunk 1 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs:157]: It's actually from BIP-21 ( the original Bitcoin URI scheme ). BIP-321 is just the modern version of BIP-21

File: components/payment_uri/tests/parse.rs [hunk 4]
```diff
@@ -0,0 +1,215 @@
+use payment_uri::{Network, PaymentRequest, SolanaRequest};
+
+#[test]
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs:11]: Are there existing test vectors for each case? Valid and invalid cases. I guess an AI could find it by giving this comment.

File: components/payment_uri/tests/parse.rs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    assert_eq!(request.link(), "https://example.com/solana-pay?order=12345");
+    assert!(PaymentRequest::parse("solana:https://example.com/solana-pay?order=12345").is_err());
+}
+
+#[test]
+fn delegates_ethereum_to_eip681() {
+    let request =
+        PaymentRequest::parse("ethereum:0xfB6916095ca1df60bB79Ce92cE3Ea74c37c5d359?value=1e18")
+            .unwrap();
+    assert!(matches!(request, PaymentRequest::Ethereum(_)));
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs]: The values in the request could be verified, confirming the parsing went well.

File: components/payment_uri/tests/parse.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+#[test]
+fn categorizes_solana_transaction_requests_without_treating_them_as_transfers() {
+    let request =
+        PaymentRequest::parse("solana:https%3A%2F%2Fexample.com%2Fsolana-pay%3Forder%3D12345")
+            .unwrap();
+    let PaymentRequest::Solana(SolanaRequest::Transaction(request)) = request else {
+        panic!("expected Solana transaction request");
+    };
+    assert_eq!(request.link(), "https://example.com/solana-pay?order=12345");
+    assert!(PaymentRequest::parse("solana:https://example.com/solana-pay?order=12345").is_err());
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs]: Interesting: the URL MUST be escaped. As for bitcoin tests above, we might want to split valid and invalid cases.

File: components/payment_uri/src/lib.rs (hunk 2 above)
Comment (zcash-harry) [reply components/payment_uri/src/lib.rs]: added it [cae51ff](https://github.com/zcash/librustzcash/pull/2989/commits/cae51fff12945404a2b41f2208b7805262d9f1a0)

File: components/payment_uri/tests/parse.rs (hunk 3 above)
Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs:13]: addressed - [c153fe7](https://github.com/zcash/librustzcash/pull/2989/commits/c153fe7c6c047b07bb0f8ccc3cf0ffb0efd78573)

File: components/payment_uri/src/bitcoin.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
+use std::collections::HashSet;
```
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs:19]: Nit: could we have a top-level doc with a link to the actual spec?
This might need to be updated over time when the different supported networks gets updates.

This applies to all parsing files. Actual spec might be useful to extract corner cases, test vectors, etc. A deep search might help.

File: components/payment_uri/src/bitcoin.rs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    pub fn label(&self) -> Option<&str> {
+        self.label.as_deref()
+    }
+
+    /// Returns the decoded payment message, if provided.
+    pub fn message(&self) -> Option<&str> {
+        self.message.as_deref()
+    }
+}
+
```
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs:120]: Are we used in the codebase to add documentation for `pub(crate)` methods? For this kind of parsing methods, I think it would be useful to have examples of inputs the method should accept, as it might ease the human reviewer to get the context.

No change requested if it is _not_ the convention in the codebase. If it is somehow a convention, this is purely a suggestion.

File: components/payment_uri/src/bitcoin.rs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        let (raw_key, raw_value) = parameter.split_once('=').unwrap_or((parameter, ""));
+        let normalized = match currency {
+            Currency::Bitcoin => raw_key.to_ascii_lowercase(),
+            Currency::Litecoin => raw_key.to_owned(),
+        };
+        let (required, key) = normalized
+            .strip_prefix("req-")
+            .map_or((false, normalized.as_str()), |key| (true, key));
+
+        if matches!(key, "amount" | "label" | "message") && !seen.insert(key.to_owned()) {
```
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs:159]: Are these all valid keys from the spec?

File: components/payment_uri/src/bitcoin.rs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    let decoded = bs58::decode(address)
+        .with_check(None)
+        .into_vec()
+        .map_err(|_| Error::InvalidAddress(address.to_owned()))?;
+    let [version, payload @ ..] = decoded.as_slice() else {
+        return Err(Error::InvalidAddress(address.to_owned()));
+    };
+    if payload.len() != 20 {
```
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs]: Very nit and personal taste (repository convention to follow first): I don't like having magic numbers in the code. If it is a constant, I like having a `const` value at the top of the file which bind the value to an explicit variable name, with its role (like `FOO_BAR_LEN` for a length).

File: components/payment_uri/src/bitcoin.rs [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        return Err(Error::InvalidAddress(address.to_owned()));
+    }
+
+    match (currency, *version) {
+        (Currency::Bitcoin, 0 | 5) | (Currency::Litecoin, 5 | 48 | 50) => Ok(Network::Mainnet),
```
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs]: Same here for the magic numbers. Linking to spec might be useful, with meaningful binding names.
Repository conventions to follow first.

File: components/payment_uri/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+/// Version of the JSON representation returned by [`parse_to_json`].
+pub const JSON_VERSION: u8 = 1;
+
+/// A non-negative decimal amount represented without floating-point loss.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DecimalAmount(String);
+
+impl DecimalAmount {
+    fn parse(value: &str) -> Result<Self, Error> {
+        let (whole, fraction) = value
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs:28]: Can we have valid and invalid tests for this? With corner cases, and adversarial values.

File: components/payment_uri/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+impl PaymentRequest {
+    /// Parses and validates a supported payment request URI.
+    pub fn parse(input: &str) -> Result<Self, Error> {
+        let (scheme, payload) = input.split_once(':').ok_or(Error::MissingScheme)?;
+        if scheme.eq_ignore_ascii_case("bitcoin") {
+            bitcoin::parse(input, bitcoin::Currency::Bitcoin).map(Self::Bitcoin)
+        } else if scheme.eq_ignore_ascii_case("litecoin") {
+            bitcoin::parse(input, bitcoin::Currency::Litecoin).map(Self::Litecoin)
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs:93]: Very nit: should we have a file `litecoin` which simply creates direct aliases? No strong opinion at all.

File: components/payment_uri/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+                "type": "solana_transaction",
+                "link": request.link(),
+            }),
+        }
+    }
+}
+
+/// Parses a payment request and returns its versioned JSON representation.
+#[cfg(feature = "json")]
+pub fn parse_to_json(input: &str) -> Result<String, Error> {
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs:133]: Would be nice to have rountrip tests, with proptest.

File: components/payment_uri/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+            "version": JSON_VERSION,
+            "type": "ethereum_unrecognised",
+        }),
+    }
+}
+
+/// A payment request parsing or validation error.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs:184]: AI question: do we have a test for each failure cases? Are we testing with proptest/corner cases?

File: components/payment_uri/src/solana.rs [hunk 16]
```diff
@@ -0,0 +1,168 @@
+use std::collections::HashSet;
+
+use crate::{DecimalAmount, Error, decode};
+
+const SOL_DECIMAL_PLACES: usize = 9;
+const PUBLIC_KEY_BYTES: usize = 32;
```
Comment (dannywillems) [inline components/payment_uri/src/solana.rs]: Nit: this represents a length, therefore it should be suffixed with its function, e.g. `PUBLIC_KEY_BYTES_LEN` or `PUBLIC_KEY_BYTES_LENGTH`.

File: components/payment_uri/Cargo.toml [hunk 17]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "payment_uri"
+description = "Parsing and validation for cross-chain payment request URIs"
+version = "0.1.0"
+authors = ["Harry <110472914+Cosmos-Harry@users.noreply.github.com>"]
```
Comment (dannywillems) [inline components/payment_uri/Cargo.toml:5]: Is it the expected email?

Comment (zcash-harry) [reply components/payment_uri/Cargo.toml:5]: thats the actual no-reply email yes ( it somehow stays the same even though username was changed ) 
You think just a regular email should be fine here?

File: components/payment_uri/src/solana.rs [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+    let mut spl_token = None;
+    let mut references = vec![];
+    let mut label = None;
+    let mut message = None;
+    let mut memo = None;
+    let mut seen = HashSet::new();
+
+    for parameter in query.into_iter().flat_map(|query| query.split('&')) {
+        let (key, value) = parameter.split_once('=').unwrap_or((parameter, ""));
+        if matches!(key, "amount" | "spl-token" | "label" | "message" | "memo") && !seen.insert(key)
```
Comment (dannywillems) [inline components/payment_uri/src/solana.rs:143]: Is it case sensitive or not? Same question for all files.

File: components/payment_uri/tests/parse.rs [hunk 19]
```diff
@@ -12,11 +12,21 @@ fn parses_bitcoin_request_and_rejects_unsafe_extensions() {
...
+fn rejects_unsupported_required_bitcoin_extension() {
     assert!(
         PaymentRequest::parse("bitcoin:1FsSia9rv4NeEwvJ2GvXrX7LyxYspbN2mo?req-unknown=true")
             .is_err()
     );
+}
+
+#[test]
+fn rejects_duplicate_bitcoin_amount_parameter() {
+    // Bitcoin (BIP-321) parameter keys are case-insensitive, so "AMOUNT" collides with "amount"
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs:35]: For each rejection, we should test which error is actually returned, and the error message should match the expected behavior.

File: components/payment_uri/tests/parse.rs [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+#[test]
+fn rejects_bitcoin_address_with_invalid_characters() {
+    // '0', 'O', 'I', 'l' are excluded from the base58 alphabet.
+    assert!(PaymentRequest::parse("bitcoin:0IOl0000000000000000000000000000").is_err());
+}
+
+#[test]
+fn rejects_address_from_the_wrong_currency() {
+    // A real Litecoin address (version byte 48) is not a valid Bitcoin address.
+    assert!(PaymentRequest::parse("bitcoin:LT2KVaAy1ppRuxRgrS5RNU3vBsy7RibPeA?amount=1").is_err());
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs]: For each error, we should explicitly test the error that is raised, with the actual messages, if any. This applies to all test files.

File: components/payment_uri/src/bitcoin.rs (hunk 11 above)
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs]: For this, I would say that it is not nitpicking to have constant value at the top-level. These magic numbers are important and are part of specs.

File: components/payment_uri/src/solana.rs [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+impl SolanaTransactionRequest {
+    /// Returns the decoded absolute HTTPS endpoint.
+    pub fn link(&self) -> &str {
+        &self.link
+    }
+}
+
+pub(crate) fn parse(input: &str) -> Result<SolanaRequest, Error> {
```
Comment (dannywillems) [inline components/payment_uri/src/solana.rs:116]: As for the comments on the other `parse`: can we stress test this method? It can be error prone. Edge cases can be existing, and as it parses URL, it is relatively important. utf characters for instance, `\`, etc. An AI agent should be able to deepthink about potential important cases.

**Review by dannywillems — COMMENTED** [review]
General comments:

Be adversarial in testing.
Link to specs.
Deep search about existing test vectors in specs and edge cases (rejecting/valid cases).

File: components/payment_uri/Cargo.toml (hunk 17 above)
Comment (dannywillems) [reply components/payment_uri/Cargo.toml:5]: > You think just a regular email should be fine here?

I would delegate this question to @nuttycom for conventions.

Comment (nuttycom) [reply components/payment_uri/Cargo.toml:5]: I use a personal email (same as my commits are attributed to) but it's a personal preference. It should resolve to an email you can actually receive at, if you specify an email at all.

File: components/payment_uri/src/lib.rs (hunk 13 above)
Comment (zcash-harry) [reply components/payment_uri/src/lib.rs:93]: How about call the current file utxo.rs?

File: components/payment_uri/tests/parse.rs (hunk 5 above)
Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs]: It now unwraps into the native ETH transfer and asserts the actual decoded fields: recipient_address equals the address from the URI.

File: components/payment_uri/tests/parse.rs (hunk 6 above)
Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs]: split into parses_solana_transaction_link / rejects_unescaped_solana_transaction_link_with_query_string

File: components/payment_uri/src/bitcoin.rs (hunk 7 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs:19]: added BIP-21/321 links and Solana Pay spec links

File: components/payment_uri/src/bitcoin.rs (hunk 8 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs:120]: It was not a convention observed but still added some examples

File: components/payment_uri/src/bitcoin.rs (hunk 9 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs:159]: They are from BIP-321's other extensions (lightning, lno, pay, sp, bc/tb, pop)

File: components/payment_uri/src/bitcoin.rs (hunk 10 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs]: added constant

File: components/payment_uri/src/bitcoin.rs (hunk 11 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs]: Constants added with doc comments

File: components/payment_uri/src/lib.rs (hunk 12 above)
Comment (zcash-harry) [reply components/payment_uri/src/lib.rs:28]: Added an inline unit test

File: components/payment_uri/src/lib.rs (hunk 14 above)
Comment (zcash-harry) [reply components/payment_uri/src/lib.rs:133]: added proptest as a dev-dependency

File: components/payment_uri/src/lib.rs (hunk 15 above)
Comment (zcash-harry) [reply components/payment_uri/src/lib.rs:184]: Every error variant has a dedicated test now

File: components/payment_uri/src/solana.rs (hunk 16 above)
Comment (zcash-harry) [reply components/payment_uri/src/solana.rs]: fixed

File: components/payment_uri/Cargo.toml (hunk 17 above)
Comment (zcash-harry) [reply components/payment_uri/Cargo.toml:5]: Added personal e-mail of this GH

File: components/payment_uri/src/solana.rs (hunk 18 above)
Comment (zcash-harry) [reply components/payment_uri/src/solana.rs:143]: Specs dont mention anything so had it case sensitive

File: components/payment_uri/tests/parse.rs (hunk 19 above)
Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs:35]: Added error messages

File: components/payment_uri/tests/parse.rs (hunk 20 above)
Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs]: Added error messages

File: components/payment_uri/src/bitcoin.rs (hunk 11 above)
Comment (zcash-harry) [reply components/payment_uri/src/bitcoin.rs]: added constants

File: components/payment_uri/src/solana.rs (hunk 21 above)
Comment (zcash-harry) [reply components/payment_uri/src/solana.rs:116]: AI : expanded the adversarial-input test with backslash/unicode/embedded-NUL/oversized-payload cases, plus a new invalid-public-key test.

File: components/payment_uri/src/bitcoin.rs [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+//! also defines a number of alternate-payment-method and privacy extensions this module does
+//! not implement -- `lightning` (BOLT 11 invoices), `lno` (BOLT 12 offers), `pay` (BIP 351
+//! private payments), `sp` (BIP 352 silent payments), `bc`/`tb` (segwit fallback addresses),
+//! and `pop`/`req-pop` (proof-of-payment callbacks). None of those are supported, but per the
+//! `req-` convention both BIPs define, an unsupported parameter is only fatal to the request
+//! when it is marked required (`req-lno=...`); as a plain optional parameter (`lightning=...`)
+//! it is safely ignored, the same as any other wallet that hasn't implemented that extension.
+//!
+//! [BIP 21]: https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki
+//! [BIP 321]: https://github.com/bitcoin/bips/blob/master/bip-0321.mediawiki
```
Comment (dannywillems) [inline components/payment_uri/src/bitcoin.rs:17]: Follow-up: we should ensure that we have a link checker in the repository.

File: components/payment_uri/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+            };
+            let parsed = DecimalAmount::parse(&text).expect("well-formed by construction");
+            let expected = whole * 10u64.pow(decimal_places) + fraction * 10u64.pow((decimal_places as usize - fraction_digits) as u32);
+            prop_assert_eq!(parsed.atomic_value(decimal_places as usize), Some(expected));
+        }
+
+        /// Any string containing a byte outside `[0-9.]` must be rejected -- parsing must never
+        /// panic on arbitrary bytes, including non-UTF-8-adjacent and adversarial input.
+        #[test]
+        fn decimal_amount_never_panics_on_arbitrary_strings(s in ".*") {
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs]: I see. It should return an error then?

File: components/payment_uri/src/solana.rs [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+    }
+}
+
+/// Parses and validates a `solana:` transfer or interactive transaction request.
+///
+/// Examples of accepted input:
+/// - `solana:mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN?amount=1&label=Michael` (native SOL)
+/// - `solana:mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN?amount=0.01&spl-token=EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v`
+///   (SPL token transfer)
+/// - `solana:https://example.com/solana-pay` (interactive request, no query string, no
```
Comment (dannywillems) [inline components/payment_uri/src/solana.rs:106]: Ah, I thought it was a rejected input. This seems different than before.

File: components/payment_uri/tests/parse.rs [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+        "ethereum:\u{0}",
+        &"bitcoin:1FsSia9rv4NeEwvJ2GvXrX7LyxYspbN2mo?amount=".repeat(10_000),
+        &format!("solana:{}", "a".repeat(100_000)),
+        "🦀:not-a-real-scheme",
+        "bitcoin://1FsSia9rv4NeEwvJ2GvXrX7LyxYspbN2mo",
+        ":",
+        "::::",
+    ];
+    for input in adversarial {
+        // The only contract under test is "does not panic"; every one of these is expected to
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs]: Maybe check that it actually raises an error? No need to verify which one.

**Review by dannywillems — DISMISSED** [review]
utACK: good shape for a first version.

File: components/payment_uri/src/lib.rs (hunk 13 above)
Comment (dannywillems) [reply components/payment_uri/src/lib.rs:93]: I don't know. `utxo` would also include Cardano and other UTXO based state managements. We can leave it like this; as said, no strong opinion at all.

File: components/payment_uri/src/lib.rs [hunk 26]
```diff
@@ -361,11 +361,22 @@ mod tests {
...
+        /// one or more digits, nothing else. Since `s` is any string at all, most generated
+        /// cases are malformed and must be rejected, but proptest can and does occasionally
+        /// generate a well-formed one (e.g. plain digit strings) by chance -- those must still
+        /// be *accepted*, so the assertion has to check both directions rather than assuming
+        /// arbitrary input is always invalid.
         #[test]
-        fn decimal_amount_never_panics_on_arbitrary_strings(s in ".*") {
-            let _ = DecimalAmount::parse(&s);
+        fn decimal_amount_matches_its_documented_grammar_for_arbitrary_strings(s in ".*") {
+            let well_formed = {
```
Comment (dannywillems) [inline components/payment_uri/src/lib.rs]: This is exactly the reimplementation of decimal::parse. Maybe my previous comment was wrong and should have been ignored.

File: components/payment_uri/tests/parse.rs [hunk 27]
```diff
@@ -334,24 +334,44 @@ fn does_not_panic_on_adversarial_input() {
...
+            "expected {input:?} to be rejected"
+        );
     }
+
+    // Two separate large-input cases, checked only for non-panic/non-hang: a query string
+    // this pathological (no '&' separators at all, just one huge repeated key=value blob)
+    // and an oversized single-field payload aren't values a fixed always_invalid list can
+    // assert a single Err/Ok answer for as confidently as the hand-picked cases above, but
+    // they must not panic or take unreasonable time either way.
+    let _ =
```
Comment (dannywillems) [inline components/payment_uri/tests/parse.rs]: So, this is a valid input?

Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs]: changed both to assert!(...is_err()). ( checked they always error )

Comment (zcash-harry) [reply components/payment_uri/tests/parse.rs]: they were both fixed strings

**Review by dannywillems — APPROVED** [review]
utACK 6d45d723c97edd4308394eaa7f3c52470a307241
